ndisasm.1
'\" t
.\"     Title: ndisasm
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 04/20/2013
.\"    Manual: The Netwide Assembler Project
.\"    Source: NASM
.\"  Language: English
.\"
.TH "NDISASM" "1" "04/20/2013" "NASM" "The Netwide Assembler Project"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndisasm \- the Netwide Disassembler, an 80x86 binary file disassembler
.SH "SYNOPSIS"
.sp
\fBndisasm\fR [ \fB\-o\fR origin ] [ \fB\-s\fR sync\-point [\&...]] [ \fB\-a\fR | \fB\-i\fR ] [ \fB\-b\fR bits ] [ \fB\-u\fR ] [ \fB\-e\fR hdrlen ] [ \fB\-p\fR vendor ] [ \fB\-k\fR offset,length [\&...]] infile
.SH "DESCRIPTION"
.sp
The \fBndisasm\fR command generates a disassembly listing of the binary file infile and directs it to stdout\&.
.SH "OPTIONS"
.PP
\fB\-h\fR
.RS 4
Causes
\fBndisasm\fR
to exit immediately, after giving a summary of its invocation options\&.
.RE
.PP
\fB\-r\fR|\fB\-v\fR
.RS 4
Causes
\fBndisasm\fR
to exit immediately, after displaying its version number\&.
.RE
.PP
\fB\-o\fR \fIorigin\fR
.RS 4
Specifies the notional load address for the file\&. This option causes
\fBndisasm\fR
to get the addresses it lists down the left hand margin, and the target addresses of PC\-relative jumps and calls, right\&.
.RE
.PP
\fB\-s\fR \fIsync\-point\fR
.RS 4
Manually specifies a synchronisation address, such that
\fBndisasm\fR
will not output any machine instruction which encompasses bytes on both sides of the address\&. Hence the instruction which starts at that address will be correctly disassembled\&.
.RE
.PP
\fB\-e\fR \fIhdrlen\fR
.RS 4
Specifies a number of bytes to discard from the beginning of the file before starting disassembly\&. This does not count towards the calculation of the disassembly offset: the first
\fIdisassembled\fR
instruction will be shown starting at the given load address\&.
.RE
.PP
\fB\-k\fR \fIoffset,length\fR
.RS 4
Specifies that
\fIlength\fR
bytes, starting from disassembly offset
\fIoffset\fR, should be skipped over without generating any output\&. The skipped bytes still count towards the calculation of the disassembly offset\&.
.RE
.PP
\fB\-a\fR|\fB\-i\fR
.RS 4
Enables automatic (or intelligent) sync mode, in which
\fBndisasm\fR
will attempt to guess where synchronisation should be performed, by means of examining the target addresses of the relative jumps and calls it disassembles\&.
.RE
.PP
\fB\-b\fR \fIbits\fR
.RS 4
Specifies 16\-, 32\- or 64\-bit mode\&. The default is 16\-bit mode\&.
.RE
.PP
\fB\-u\fR
.RS 4
Specifies 32\-bit mode, more compactly than using \(oq\-b 32\(cq\&.
.RE
.PP
\fB\-p\fR \fIvendor\fR
.RS 4
Prefers instructions as defined by
\fIvendor\fR
in case of a conflict\&. Known
\fIvendor\fR
names include
\fBintel\fR,
\fBamd\fR,
\fBcyrix\fR, and
\fBidt\fR\&. The default is
\fBintel\fR\&.
.RE
.SH "RESTRICTIONS"
.sp
\fBndisasm\fR only disassembles binary files: it has no understanding of the header information present in object or executable files\&. If you want to disassemble an object file, you should probably be using \fBobjdump\fR(1)\&.
.sp
Auto\-sync mode won\(cqt necessarily cure all your synchronisation problems: a sync marker can only be placed automatically if a jump or call instruction is found to refer to it \fIbefore\fR \fBndisasm\fR actually disassembles that part of the code\&. Also, if spurious jumps or calls result from disassembling non\-machine\-code data, sync markers may get placed in strange places\&. Feel free to turn auto\-sync off and go back to doing it manually if necessary\&.
.SH "SEE ALSO"
.sp
\fBobjdump\fR(1)


CHANGES
2.04
----
* Sanitize macro handing in the %error directive.
* New %warning preprocessor directive.
* __utf16__ and __utf32__ operators to generate UTF-16 and UTF-32
  strings.
* Fix bug in case-insensitive matching when compiled on platforms that
  don't use the "configure" script.  Of the official release binaries,
  that only affects the OS/2 binary.
* Correct the handling of nested %reps.
* Support for x87 packed BCD constants.
* New %warning directive to issue user-controlled warnings.
* New %strcat directive to join strings together.
* Correct the LTR instruction in 64-bit mode.
* Fix unnecessary REX.W prefix on indirect jumps in 64-bit mode.

2.03.01
-------
* Fix buffer overflow in the listing module.
* Fix the handling of hexadecimal escape codes in `...` strings.
* The Postscript/PDF documentation has been reformatted.
* The -F option now implies -g.

2.03
----
* Add support for Intel AVX, CLMUL and FMA instructions,
  including YMM registers.
* "dy", "resy" and "yword" for 32-byte operands.
* Fix some SSE5 instructions.
* Intel INVEPT, INVVPID and MOVBE instructions.
* Fix checking for critical expressions when the optimizer is enabled.
* Support the DWARF debugging format for ELF targets.
* Fix optimizations of signed bytes.
* Fix operation on bigendian machines.
* Fix buffer overflow in the preprocessor.
* SAFESEH support for Win32, IMAGEREL for Win64 (SEH).
* %? and %?? to refer to the name of a macro itself.  In particular,
  "%idefine keyword $%?" can be used to make a keyword "disappear".
* New options for dependency generation: -MD, -MF, -MP, -MT, -MQ.
* New preprocessor directives %pathsearch and %depend; INCBIN
  reimplemented as a macro.
* %include now resolves macros in a sane manner.
* %substr can now be used to get other than one-character substrings.
* New type of character/string constants, using backquotes (`...`),
  which support C-style escape sequences.
* %defstr and %idefstr to stringize macro definitions before creation.
* Fix forward references used in EQU statements.

2.02
----
* Additional fixes for MMX operands with explicit "qword", as well as
  (hopefully) SSE operands with "oword".
* Fix handling of truncated strings with DO.
* Fix segfaults due to memory overwrites when floating-point constants
  were used.
* Fix segfaults due to missing include files.
* Fix OpenWatcom Makefiles for DOS and OS/2.
* Add autogenerated instruction list back into the documentation.
* ELF: Fix segfault when generating stabs, and no symbols have been
  defined.
* ELF: Experimental support for DWARF debugging information.
* New compile date and time standard macros.
* %ifnum now returns true for negative numbers.
* New %iftoken test for a single token.
* New %ifempty test for empty expansion.
* Add support for the XSAVE instruction group.
* Makefile for Netware/gcc.
* Fix issue with some warnings getting emitted way too many times.
* Autogenerated instruction list added to the documentation.

2.01
----
* Fix the handling of MMX registers with explicit "qword" tags on
  memory (broken in 2.00 due to 64-bit changes.)
* Fix the PREFETCH instructions.
* Fix the documentation.
* Fix debugging info when using "-f elf" (backwards alias for "-f
  elf32").
* Man pages for rdoff tools (from the Debian project.)
* ELF: handle large numbers of sections.
* Fix corrupt output when the optimizer runs out of passes.

2.00
----
* Added c99 data-type compliance.
* Added general x86-64 support.
* Added win64 (x86-64 COFF) output format.
* Added __BITS__ standard macro.
* Renamed the elf output format to elf32 for clarity.
* Added elf64 and macho (MacOS X) output formats.
* Added Numeric constants in DQ directive.
* Added oword, do and reso pseudo operands.
* Allow underscores in numbers.
* Added 8-, 16- and 128-bit floating-point formats.
* Added binary, octal and hexadecimal floating-point.
* Correct the generation of floating-point constants.
* Added Floating-point option control.
* Added Infinity and NaN floating point support.
* Added ELF Symbol Visibility support.
* Added Setting OSABI value in ELF header directive.
* Added Generate Makefile Dependencies option.
* Added Unlimited Optimization Passes option.
* Added %IFN and %ELIFN support.
* Added Logical Negation Operator.
* Enhanced Stack Relative Preprocessor Directives.
* Enhanced ELF Debug Formats.
* Enhanced Send Errors to a File option.
* Added SSSE3, SSE4.1, SSE4.2, SSE5 support.
* Added a large number of additional instructions.
* Significant performance improvements.

0.98.39
-------
* fix buffer overflow
* fix outas86's .bss handling
* "make spotless" no longer deletes config.h.in.
* %(el)if(n)idn insensitivity to string quotes difference (#809300).
* (nasm.c) __OUTPUT_FORMAT__ changed to string value instead of symbol.

0.98.38
-------
* Add Makefile for 16-bit DOS binaries under OpenWatcom, and modify
  mkdep.pl to be able to generate completely pathless dependencies, as
  required by OpenWatcom wmake (it supports path searches, but not
  explicit paths.)
* Fix the STR instruction.
* Fix the ELF output format, which was broken under certain
  circumstances due to the addition of stabs support.
* Quick-fix Borland format debug-info for -f obj  
* Fix for %rep with no arguments (#560568)
* Fix concatenation of preprocessor function call (#794686)
* Fix long label causes coredump (#677841)
* Use autoheader as well as autoconf to keep configure from generating
  ridiculously long command lines.
* Make sure that all of the formats which support debugging output
  actually will suppress debugging output when -g not specified.

0.98.37
-------
* Paths given in "-I" switch searched for "incbin"ed as
  well as "%include"ed files.
* Added stabs debugging for the ELF output format, patch from
  Martin Wawro.
* Fix output/outbin.c to allow origin > 80000000h.
* Make -U switch work.
* Fix the use of relative offsets with explicit prefixes, e.g. "a32 loop foo".
* Remove "backslash()".
* Fix the SMSW and SLDT instructions.

ndisasm.c
/* ----------------------------------------------------------------------- *
 *   
 *   Copyright 1996-2009 The NASM Authors - All Rights Reserved
 *   See the file AUTHORS included with the NASM distribution for
 *   the specific copyright holders.
 *
 *   Redistribution and use in source and binary forms, with or without
 *   modification, are permitted provided that the following
 *   conditions are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *     
 *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *     CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *     INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *     MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *     DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *     SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *     NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *     LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *     HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *     CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 *     OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 *     EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ----------------------------------------------------------------------- */

/*
 * ndisasm.c   the Netwide Disassembler main module
 */

#include "compiler.h"

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <inttypes.h>

#include "insns.h"
#include "nasm.h"
#include "nasmlib.h"
#include "sync.h"
#include "disasm.h"

#define BPL 8                   /* bytes per line of hex dump */

static const char *help =
    "usage: ndisasm [-a] [-i] [-h] [-r] [-u] [-b bits] [-o origin] [-s sync...]\n"
    "               [-e bytes] [-k start,bytes] [-p vendor] file\n"
    "   -a or -i activates auto (intelligent) sync\n"
    "   -u same as -b 32\n"
    "   -b 16, -b 32 or -b 64 sets the processor mode\n"
    "   -h displays this text\n"
    "   -r or -v displays the version number\n"
    "   -e skips <bytes> bytes of header\n"
    "   -k avoids disassembling <bytes> bytes from position <start>\n"
    "   -p selects the preferred vendor instruction set (intel, amd, cyrix, idt)\n";

static void output_ins(uint32_t, uint8_t *, int, char *);
static void skip(uint32_t dist, FILE * fp);

static void ndisasm_verror(int severity, const char *fmt, va_list va)
{
    vfprintf(stderr, fmt, va);

    if (severity & ERR_FATAL)
	exit(1);
}

int main(int argc, char **argv)
{
    char buffer[INSN_MAX * 2], *p, *ep, *q;
    char outbuf[256];
    char *pname = *argv;
    char *filename = NULL;
    uint32_t nextsync, synclen, initskip = 0L;
    int lenread;
    int32_t lendis;
    bool autosync = false;
    int bits = 16, b;
    bool eof = false;
    uint32_t prefer = 0;
    bool rn_error;
    int32_t offset;
    FILE *fp;

    tolower_init();
    nasm_set_verror(ndisasm_verror);
    nasm_init_malloc_error();

    offset = 0;
    init_sync();

    while (--argc) {
        char *v, *vv, *p = *++argv;
        if (*p == '-' && p[1]) {
            p++;
            while (*p)
                switch (nasm_tolower(*p)) {
                case 'a':      /* auto or intelligent sync */
                case 'i':
                    autosync = true;
                    p++;
                    break;
                case 'h':
                    fputs(help, stderr);
                    return 0;
                case 'r':
                case 'v':
                    fprintf(stderr,
                            "NDISASM version %s compiled on %s\n",
			    nasm_version, nasm_date);
                    return 0;
                case 'u':	/* -u for -b 32, -uu for -b 64 */
		    if (bits < 64)
			bits <<= 1;
                    p++;
                    break;
                case 'b':      /* bits */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-b' requires an argument\n",
                                pname);
                        return 1;
                    }
		    b = strtoul(v, &ep, 10);
		    if (*ep || !(bits == 16 || bits == 32 || bits == 64)) {
                        fprintf(stderr, "%s: argument to `-b' should"
                                " be 16, 32 or 64\n", pname);
                    } else {
			bits = b;
		    }
                    p = "";     /* force to next argument */
                    break;
                case 'o':      /* origin */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-o' requires an argument\n",
                                pname);
                        return 1;
                    }
                    offset = readnum(v, &rn_error);
                    if (rn_error) {
                        fprintf(stderr,
                                "%s: `-o' requires a numeric argument\n",
                                pname);
                        return 1;
                    }
                    p = "";     /* force to next argument */
                    break;
                case 's':      /* sync point */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-s' requires an argument\n",
                                pname);
                        return 1;
                    }
                    add_sync(readnum(v, &rn_error), 0L);
                    if (rn_error) {
                        fprintf(stderr,
                                "%s: `-s' requires a numeric argument\n",
                                pname);
                        return 1;
                    }
                    p = "";     /* force to next argument */
                    break;
                case 'e':      /* skip a header */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-e' requires an argument\n",
                                pname);
                        return 1;
                    }
                    initskip = readnum(v, &rn_error);
                    if (rn_error) {
                        fprintf(stderr,
                                "%s: `-e' requires a numeric argument\n",
                                pname);
                        return 1;
                    }
                    p = "";     /* force to next argument */
                    break;
                case 'k':      /* skip a region */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-k' requires an argument\n",
                                pname);
                        return 1;
                    }
                    vv = strchr(v, ',');
                    if (!vv) {
                        fprintf(stderr,
                                "%s: `-k' requires two numbers separated"
                                " by a comma\n", pname);
                        return 1;
                    }
                    *vv++ = '\0';
                    nextsync = readnum(v, &rn_error);
                    if (rn_error) {
                        fprintf(stderr,
                                "%s: `-k' requires numeric arguments\n",
                                pname);
                        return 1;
                    }
                    synclen = readnum(vv, &rn_error);
                    if (rn_error) {
                        fprintf(stderr,
                                "%s: `-k' requires numeric arguments\n",
                                pname);
                        return 1;
                    }
                    add_sync(nextsync, synclen);
                    p = "";     /* force to next argument */
                    break;
                case 'p':      /* preferred vendor */
                    v = p[1] ? p + 1 : --argc ? *++argv : NULL;
                    if (!v) {
                        fprintf(stderr, "%s: `-p' requires an argument\n",
                                pname);
                        return 1;
                    }
                    if (!strcmp(v, "intel")) {
                        prefer = 0;     /* Default */
                    } else if (!strcmp(v, "amd")) {
                        prefer = IF_AMD | IF_3DNOW;
                    } else if (!strcmp(v, "cyrix")) {
                        prefer = IF_CYRIX | IF_3DNOW;
                    } else if (!strcmp(v, "idt") || !strcmp(v, "centaur")
                               || !strcmp(v, "winchip")) {
                        prefer = IF_3DNOW;
                    } else {
                        fprintf(stderr,
                                "%s: unknown vendor `%s' specified with `-p'\n",
                                pname, v);
                        return 1;
                    }
                    p = "";     /* force to next argument */
                    break;
                default:       /*bf */
                    fprintf(stderr, "%s: unrecognised option `-%c'\n",
                            pname, *p);
                    return 1;
                }
        } else if (!filename) {
            filename = p;
        } else {
            fprintf(stderr, "%s: more than one filename specified\n",
                    pname);
            return 1;
        }
    }

    if (!filename) {
        fprintf(stderr, help, pname);
        return 0;
    }

    if (strcmp(filename, "-")) {
        fp = fopen(filename, "rb");
        if (!fp) {
            fprintf(stderr, "%s: unable to open `%s': %s\n",
                    pname, filename, strerror(errno));
            return 1;
        }
    } else
        fp = stdin;

    if (initskip > 0)
        skip(initskip, fp);

    /*
     * This main loop is really horrible, and wants rewriting with
     * an axe. It'll stay the way it is for a while though, until I
     * find the energy...
     */

    p = q = buffer;
    nextsync = next_sync(offset, &synclen);
    do {
        uint32_t to_read = buffer + sizeof(buffer) - p;
	if ((nextsync || synclen) &&
	    to_read > nextsync - offset - (p - q))
            to_read = nextsync - offset - (p - q);
        if (to_read) {
            lenread = fread(p, 1, to_read, fp);
            if (lenread == 0)
                eof = true;     /* help along systems with bad feof */
        } else
            lenread = 0;
        p += lenread;
        if ((nextsync || synclen) &&
	    (uint32_t)offset == nextsync) {
            if (synclen) {
                fprintf(stdout, "%08"PRIX32"  skipping 0x%"PRIX32" bytes\n",
			offset, synclen);
                offset += synclen;
                skip(synclen, fp);
            }
            p = q = buffer;
            nextsync = next_sync(offset, &synclen);
        }
        while (p > q && (p - q >= INSN_MAX || lenread == 0)) {
            lendis =
                disasm((uint8_t *) q, outbuf, sizeof(outbuf), bits,
		       offset, autosync, prefer);
            if (!lendis || lendis > (p - q)
                || ((nextsync || synclen) &&
		    (uint32_t)lendis > nextsync - offset))
                lendis = eatbyte((uint8_t *) q, outbuf, sizeof(outbuf), bits);
            output_ins(offset, (uint8_t *) q, lendis, outbuf);
            q += lendis;
            offset += lendis;
        }
        if (q >= buffer + INSN_MAX) {
            uint8_t *r = (uint8_t *) buffer, *s = (uint8_t *) q;
            int count = p - q;
            while (count--)
                *r++ = *s++;
            p -= (q - buffer);
            q = buffer;
        }
    } while (lenread > 0 || !(eof || feof(fp)));

    if (fp != stdin)
        fclose(fp);

    return 0;
}

static void output_ins(uint32_t offset, uint8_t *data,
                       int datalen, char *insn)
{
    int bytes;
    fprintf(stdout, "%08"PRIX32"  ", offset);

    bytes = 0;
    while (datalen > 0 && bytes < BPL) {
        fprintf(stdout, "%02X", *data++);
        bytes++;
        datalen--;
    }

    fprintf(stdout, "%*s%s\n", (BPL + 1 - bytes) * 2, "", insn);

    while (datalen > 0) {
        fprintf(stdout, "         -");
        bytes = 0;
        while (datalen > 0 && bytes < BPL) {
            fprintf(stdout, "%02X", *data++);
            bytes++;
            datalen--;
        }
        fprintf(stdout, "\n");
    }
}

/*
 * Skip a certain amount of data in a file, either by seeking if
 * possible, or if that fails then by reading and discarding.
 */
static void skip(uint32_t dist, FILE * fp)
{
    char buffer[256];           /* should fit on most stacks :-) */

    /*
     * Got to be careful with fseek: at least one fseek I've tried
     * doesn't approve of SEEK_CUR. So I'll use SEEK_SET and
     * ftell... horrible but apparently necessary.
     */
    if (fseek(fp, dist + ftell(fp), SEEK_SET)) {
        while (dist > 0) {
            uint32_t len = (dist < sizeof(buffer) ?
                                 dist : sizeof(buffer));
            if (fread(buffer, 1, len, fp) < len) {
                perror("fread");
                exit(1);
            }
            dist -= len;
        }
    }
}
